image/validate_seed_test.go

```diff
@@ -0,0 +1,252 @@
+}
+
+func (s *validateSuite) makeSnapInSeed(c *C, snapYaml string) {
+	info := infoFromSnapYaml(c, snapYaml, snap.R(1))
+
+	src := snaptest.MakeTestSnapWithFiles(c, snapYaml, nil)
+	dst := filepath.Join(s.root, "snaps", fmt.Sprintf("%s_%s.snap", info.InstanceName(), info.Revision.String()))
+
+	err := os.Rename(src, dst)
+	c.Assert(err, IsNil)
+}
+
+func (s *validateSuite) makeSeedYaml(c *C, seedYaml string) string {
+	tmpf := filepath.Join(s.root, "seed.yaml")
+	err := ioutil.WriteFile(tmpf, []byte(seedYaml), 0644)
+	c.Assert(err, IsNil)
+	return tmpf
+}
+
+func (s *validateSuite) TestValidateSnapHappy(c *C) {
+	s.makeSnapInSeed(c, coreYaml)
+	s.makeSnapInSeed(c, `name: gtk-common-themes
+version: 19.04`)
+	seedFn := s.makeSeedYaml(c, `
+snaps:
+ - name: core
+   channel: stable
+   file: core_1.snap
+ - name: gtk-common-themes
+   channel: stable/ubuntu-19.04
+   file: gtk-common-themes_1.snap
+`)
+
+	err := image.ValidateSeed(seedFn)
+	c.Assert(err, IsNil)
+}
+
+func (s *validateSuite) TestValidateSnapMissingBase(c *C) {
+	s.makeSnapInSeed(c, `name: need-base
+base: some-base
+version: 1.0`)
+	s.makeSnapInSeed(c, coreYaml)
+	seedFn := s.makeSeedYaml(c, `
+snaps:
+ - name: core
+   file: core_1.snap
+ - name: need-base
+   file: need-base_1.snap
+`)
+
+	err := image.ValidateSeed(seedFn)
+	c.Assert(err, ErrorMatches, `cannot validate seed:
+- cannot use snap "need-base": base "some-base" is missing`)
+}
+
+func (s *validateSuite) TestValidateSnapMissingDefaultProvider(c *C) {
+	s.makeSnapInSeed(c, coreYaml)
+	s.makeSnapInSeed(c, `name: need-df
+version: 1.0
+plugs:
+ gtk-3-themes:
+  interface: content
+  default-provider: gtk-common-themes
+`)
+	seedFn := s.makeSeedYaml(c, `
+snaps:
+ - name: core
+   file: core_1.snap
+ - name: need-df
+   file: need-df_1.snap
+`)
+
+	err := image.ValidateSeed(seedFn)
+	c.Assert(err, ErrorMatches, `cannot validate seed:
+- cannot use snap "need-df": default provider "gtk-common-themes" is missing`)
+}
+
+func (s *validateSuite) TestValidateSnapSnapdHappy(c *C) {
+	s.makeSnapInSeed(c, snapdYaml)
+	s.makeSnapInSeed(c, packageCore18)
+	s.makeSnapInSeed(c, `name: some-snap
+version: 1.0
+base: core18
+`)
+	seedFn := s.makeSeedYaml(c, `
+snaps:
+ - name: snapd
+   file: snapd_1.snap
+ - name: some-snap
+   file: some-snap_1.snap
+ - name: core18
+   file: core18_1.snap
+`)
+
+	err := image.ValidateSeed(seedFn)
+	c.Assert(err, IsNil)
+}
+
+func (s *validateSuite) TestValidateSnapMissingCore(c *C) {
+	s.makeSnapInSeed(c, snapdYaml)
+	s.makeSnapInSeed(c, `name: some-snap
+version: 1.0`)
+	seedFn := s.makeSeedYaml(c, `
+snaps:
+ - name: snapd
+   file: snapd_1.snap
+ - name: some-snap
+   file: some-snap_1.snap
+`)
+
+	err := image.ValidateSeed(seedFn)
+	c.Assert(err, ErrorMatches, `cannot validate seed:
+- cannot use snap "some-snap": required snap "core" missing`)
+}
+
+func (s *validateSuite) TestValidateSnapMissingSnapdAndCore(c *C) {
+	s.makeSnapInSeed(c, packageCore18)
+	s.makeSnapInSeed(c, `name: some-snap
```

Comment:
This function will create a snap with name `some-snap`, however it's not clear to me that will get cleaned up before the next test, which requires that `some-snap` doesn't exist in the seed.

Should there be some cleanup we do after the tests to remove the created snap files? Or is this already done automatically.

Reply:
Thanks for this careful check of the PR - it will be done automatically, makeSnapInSeed will install the snaps into "s.root" which is create in imageSuite.SetUpTest() via s.root = c.MkDir(). And c.MkDir() creates a per-test specific tempdir and also cleans this up after each test.

Reply:
Okay, thanks for clarifying